lib/shape.ml

```diff
@@ -1,5 +1,72 @@
 open Graphics
 
+type point = { x : int; y : int }
+type rectangle = { c : point; length : int; width : int }
+type circle = { c : point; radius : int }
+type ellipse = { c : point; rx : int; ry : int }
+type shape = Circle of circle | Rectangle of rectangle | Ellipse of ellipse
+type shapes = shape list
+
+let canvas_mid = { x = 250; y = 250 }
+
+let render_shape s =
+  match s with
+  | Circle circle -> draw_circle circle.c.x circle.c.y circle.radius
+  | Rectangle rectangle ->
+      draw_rect rectangle.c.x rectangle.c.y rectangle.length rectangle.width
+  | Ellipse ellipse ->
+      draw_ellipse ellipse.c.x ellipse.c.y ellipse.rx ellipse.ry
+
+let circle ?x ?y r =
+  match (x, y) with
+  | Some x, Some y -> Circle { c = { x; y }; radius = r }
+  | _ -> Circle { c = { x = canvas_mid.x; y = canvas_mid.y }; radius = r }
+
+let rectangle ?x ?y length width =
+  match (x, y) with
+  | Some x, Some y -> Rectangle { c = { x; y }; length; width }
+  | _ -> Rectangle { c = { x = canvas_mid.x; y = canvas_mid.y }; length; width }
+
+let ellipse ?x ?y rx ry =
+  match (x, y) with
+  | Some x, Some y -> Ellipse { c = { x; y }; rx; ry }
+  | _ -> Ellipse { c = { x = canvas_mid.x; y = canvas_mid.y }; rx; ry }
+
+let translate dx dy shape =
+  match shape with
+  | Circle circle -> Circle { circle with c = { x = circle.c.x + dx; y = circle.c.y + dy } }
+  | Rectangle rectangle -> Rectangle { rectangle with c = { x = rectangle.c.x + dx; y = rectangle.c.y + dy } }
+  | Ellipse ellipse -> Ellipse { ellipse with c = { x = ellipse.c.x + dx; y = ellipse.c.y + dy } }
+
+let show shapes = List.iter render_shape shapes
+
+let init () =
+  open_graph " 500x500";
+  set_color black
+
+let close () =
+  ignore (read_line ());
+  close_graph ()
+
+let () =
+  init ();
+  set_color black;
+
+  (* Create shapes *)
+  let r1 = rectangle 200 100 in
+  let r2 = translate 100 0 r1 in
+
+  (* Display shapes *)
+  show [r1; r2];
+
+  close ()
```

Comment:
Great work! 🎉

the lib is not executed by default.
as a convention is most projects, the `lib` directory is dedicated for all the *buildup* code, that is then imported into something you want to be executable.

You should instead copy this into a new file in the `examples/` directory, along with the necessary imports.

Then try building the examples directory and running that file with `dune build` and `dune exec` like mentioned in the earlier issues.

Reply:
Okay, thank you for this! I'll try that now.

Reply:
I have created new shape.ml and shape.mli files in the `examples` directory and copied the codes into them. 

Is that what you asked me to do or am I supposed to copy just some part of the code into a new file?

I've done the above and the result is still not displaying so I want to confirm if I'm doing it right.

Reply:
> Great work! 🎉
> 
> Anything in `lib/` is not executed, as a convention. In most projects, the `lib` directory is dedicated for all the _buildup_ code, that is then imported into some other module that you want to be executable.
> 
> You should instead copy this into a new file in the `examples/` directory, along with the necessary imports.
> 
> Then try building the examples directory and running that file with `dune build` and `dune exec` like mentioned in the earlier issues.

I wish to ask, what file in `examples/` should I transfer this part of the code in? And what are the necessary imports that you mentioned?

Reply:
You can call it `translate_rectangle.ml` for this example. That is specific and makes it clear what the example is about.

Reply:
> You can call it `translate_rectangle.ml` for this example. That is specific and makes it clear what the example is about.

Okay, thank you

Reply:
> You can call it `translate_rectangle.ml` for this example. That is specific and makes it clear what the example is about.

So for each shape's translate transformation, I add a new file for it?